File: docs/Software/POO/EP/ep_cpp/Bateria.cpp
#include "Bateria.hpp"
#include <iostream>
using namespace std;

Bateria::Bateria(int mah, int tempoDeCarregamento){
    this->mah = mah;
    this->tempoDeCarregamento = tempoDeCarregamento;
    this->carga = 0;
    this->carregavel = true;
    this->uso = false;
}

Bateria::~Bateria(){
    cout << "Bateria destruida com sucesso" << endl;
}

void Bateria::carregar(int tempo){
    if (carga == mah)
        cout << "Bateria ja esta carregada!" << endl;
    else {
        carga = carga + (mah * tempo) / tempoDeCarregamento;
        if(carga > mah)
            carga = mah;
        cout << "Depois de carregar por " << tempo << "tempos, a bateria ficou com " << carga << "mah" << endl;

    }
}

bool Bateria::usar(int tempo){
    if(carga == 0){
        cout << "bateria descarregada :c" << endl;
        return false;
    }
    else{
        int cargaTemp  = carga - (mah * tempo) / tempoDeCarregamento;
        if (cargaTemp < 0)
            return false;
        else{
            carga = cargaTemp;
            return true;
        }
    }
}

int Bateria::getCarga(){
    return carga;
}

int Bateria::getTempoDeCarregamento(){
    return tempoDeCarregamento;
}

int Bateria::calculaTempoDeVoo(){
    return (tempoDeCarregamento * carga)/mah;
}

void Bateria::status(){
    cout << "Especificacoes: " << endl << "\tmAh: " << mah << endl;
    cout << "\tTempo de carregamento: " << tempoDeCarregamento << endl;
    cout << "\tAtualmente com carga: " << carga << " e tempo de voo restante: " << this->calculaTempoDeVoo() << endl;
    cout << "\tÉ carregavel: " << carregavel << endl;
    cout << "\tEstá sendo usada: " << uso << endl;
}

bool Bateria::getUso(){
    return uso;
}

bool Bateria::getCarregavel(){
    return carregavel;
}

void Bateria::setUso(bool uso){
    this->uso = uso;
}

void Bateria::setCarregavel(bool carregavel){
    this->carregavel = carregavel;
}

File: docs/Software/POO/EP/ep_python/Bateria.py
class Bateria():
    def __init__(self, mah, tempoDeCarregamento):
        self._mah = mah
        self._tempoDeCarregamento = tempoDeCarregamento
        self._carga = 0
        self._uso = False
        self._carregavel = True
    
    def carregar(self, tempo):
        if self._carga == self._mah:
            print("Bateria ja esta carregada!")
            return False
        else:
            self._carga = self._carga + (self._mah * tempo) / self._tempoDeCarregamento
            if self._carga > self._mah:
                self._carga = self._mah
            print("Depois de carregar por " + str(tempo) + " tempos, a bateria ficou com " + str(self._carga) + " mAh")
            return True

    def usar(self, tempo):
        if self._carga < (self._mah * tempo) / self._tempoDeCarregamento:
            print("Bateria insuficiente :C\n")
            return False
        else:
            self._carga = self._carga - (self._mah * tempo) / self._tempoDeCarregamento
            return True

    def calculaTempoDeVoo(self):
        return (self._tempoDeCarregamento * self._carga) / self._mah 
    
    def status(self):
        print("Especificacoes:")
        print("     mAh: " + str(self._mah))
        print("     Tempo de Carregamento: " + str(self._tempoDeCarregamento))
        print("     Carga Atual: " + str(self._carga))
        print("     Tempo de Voo restante: " + str(self.calculaTempoDeVoo()))
        print("     Pode ser carregada no momento: " + str(self._carregavel))
        print("     Esta sendo usada: " + str(self._uso))